### DISCUSSION TITLE
Measuring GPOS (sub)tables file size

### DISCUSSION THREAD

**Discussion by madig** [fonttools/fonttools#2268] [Q&A]
Hi!
I want to figure out why in a particular font added kerning has an outsized impact on file size. For this, I'd like to now the byte count of the serialized (sub)table data inside the GPOS table. What's the easiest way to get at them?

Edit: this measures lookup size, but I still want to measure subtable size:

```python
import argparse

from fontTools.ttLib import TTFont
from fontTools.ttLib.tables.otBase import OTTableWriter

parser = argparse.ArgumentParser()
parser.add_argument("font", type=TTFont)
parsed_args = parser.parse_args()
font: TTFont = parsed_args.font

gpos = font["GPOS"]
for index, lookup in enumerate(gpos.table.LookupList.Lookup):
    w = OTTableWriter(tableTag=gpos.tableTag)
    lookup.compile(w, font)
    print(f"Lookup {index}: {len(w.getAllData())} bytes")
```

**Comment by behdad** [comment]
That snippet should work on almost all types in GSUB/GPOS. Does it not?

**Reply by madig** [reply]
It does, I just additionally wanted to see how large the `SubTable`s were. And what's taking up all the bytes inside. Cosimo told me that internal sharing might skew the results, though...

**Reply by behdad** [reply]
I meant, can't you compile the `SubTable`s individually as well? Or maybe I'm misunderstanding still.

Or describe what kind of data / problem you are trying to figure out?

**Reply by madig** [reply]
I can, in a way, although I'm not entirely sure how to best do that.

I'm trying to figure out why a certain font has a huge GPOS size that is ballooned further by the inclusion of a new kerned script, where removing chunks of UFO kerning makes a minimal dent. Think of it as a "WinDirStat" for OT tables: 
![image](https://user-images.githubusercontent.com/380829/114839614-e160b500-9dcd-11eb-8a79-554873c3ad9f.png)

**Reply by behdad** [reply]
I see. Yeah unfortunately we don't have something like that. The closest I can think of is a JS tool by @Pomax IIRC, this:

  https://github.com/Pomax/A-binary-parser-generator

Another place to check would be a non-free binary editor called "010 Editor". I have not tried it myself, but it has an OpenType schema:

  https://www.sweetscape.com/010editor/repository/templates/file_info.php?file=OpenType.bt

cc @simoncozens since you seem to be writing an OpenType schema now.

**Reply by moyogo** [reply]
https://github.com/caryll/otfcc gives a nice json output. I think it makes it clear when values are zeroes:
```json
                        "matrix": [
                            [0,-20,-30,-50,-40,-40,-20,-20,-30,-10,-20,-40,-50,-90,-40,-10,-20,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0
,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0
,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0
,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0
,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0
,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0
,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0
,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0
,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0],
```

**Comment by simoncozens** [comment]
Here's the relevant routine from otlBuilder:

```python
    def getCompiledSize_(self, subtables):
        size = 0
        for st in subtables:
            w = OTTableWriter()
            w["LookupType"] = CountReference(
                {"LookupType": st.LookupType}, "LookupType"
            )
            # We need to make a copy here because compiling
            # modifies the subtable (finalizing formats etc.)
            copy.deepcopy(st).compile(w, self.font)
            size += len(w.getAllData())
        return size
```

**Reply by behdad** [reply]
> ```python
>             # We need to make a copy here because compiling
>             # modifies the subtable (finalizing formats etc.)
> ```

That was recently fixed. Can you try removing the `deepcopy`?
